src/setup_payload/AdditionalDataPayloadGenerator.cpp

```diff
@@ -0,0 +1,133 @@
+                                       ArraySize(rotatingDeviceIdBuffer), rotatingDeviceIdBufferSize);
+        SuccessOrExit(err);
+
+        // Adding the rotating device id to the TLV data
+        err = innerWriter.PutString(ContextTag(kRotatingDeviceIdTag), rotatingDeviceIdBuffer);
+        SuccessOrExit(err);
+    }
+
+    err = writer.CloseContainer(innerWriter);
+    SuccessOrExit(err);
+
+    err = writer.Finalize(&bufferHandle);
+    SuccessOrExit(err);
+
+exit:
+    return err;
+}
+
+CHIP_ERROR AdditionalDataPayloadGenerator::generateRotatingDeviceId(uint16_t lifetimeCounter, const char * serialNumberBuffer,
+                                                                    size_t serialNumberBufferSize, char rotatingDeviceIdBuffer[],
+                                                                    size_t rotatingDeviceIdBufferSize,
+                                                                    size_t & rotatingDeviceIdBufferOutputSize)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    Hash_SHA256_stream hash;
+    uint8_t outputBuffer[RotatingDeviceId::kRotatingDeviceIdMaxLength];
+    uint8_t hashOutputBuffer[kSHA256_Hash_Length];
+    BufferWriter outputBufferWriter(&outputBuffer[0], ArraySize(outputBuffer));
+    size_t rotatingDeviceIdBufferIndex = 0;
+    const PacketBufferHandle & lifetimeCounterBufferHandle =
+        chip::System::PacketBufferHandle::New(chip::System::PacketBuffer::kMaxSize);
+    uint8_t * lifetimeCounterBuffer = lifetimeCounterBufferHandle->Start();
+
+    VerifyOrExit(rotatingDeviceIdBufferSize >= RotatingDeviceId::kRotatingDeviceIdHexMaxLength, err = CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    chip::Encoding::LittleEndian::Write16(lifetimeCounterBuffer, lifetimeCounter);
+
+    // Computing the Rotating Device Id
+    // RDI = Lifetime_Counter + SuffixBytes(SHA256(Serial_Number + Lifetime_Counter), 16)
+
+    err = hash.Begin();
+    SuccessOrExit(err);
+
+    err = hash.AddData(Uint8::from_const_char(serialNumberBuffer), serialNumberBufferSize);
+    SuccessOrExit(err);
+
+    err = hash.AddData(lifetimeCounterBuffer, sizeof(lifetimeCounter));
+    SuccessOrExit(err);
+    err = hash.Finish(hashOutputBuffer);
+    SuccessOrExit(err);
+
+    outputBufferWriter.Put16(lifetimeCounter);
+    outputBufferWriter.Put(&hashOutputBuffer[kSHA256_Hash_Length - RotatingDeviceId::kRotatingDeviceIdHashSuffixLength],
+                           RotatingDeviceId::kRotatingDeviceIdHashSuffixLength);
+
+    for (; rotatingDeviceIdBufferIndex < outputBufferWriter.Needed(); rotatingDeviceIdBufferIndex++)
+    {
+        snprintf(&rotatingDeviceIdBuffer[rotatingDeviceIdBufferIndex * 2],
+                 rotatingDeviceIdBufferSize - rotatingDeviceIdBufferIndex * 2, "%02X",
+                 outputBufferWriter.Buffer()[rotatingDeviceIdBufferIndex]);
+    }
+
+    rotatingDeviceIdBuffer[rotatingDeviceIdBufferIndex * 2 + 1] = 0;
```

Comment:
This is off by one.  Should not have the `+1` here.

Reply:
i guess you're referring to the same line, you're correct, i will omit the +1